dump.go

```diff
@@ -0,0 +1,110 @@
+package jet
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+)
+
+// dumpAll returns
+//  - everything in Runtime.context
+//  - everything in Runtime.variables
+//  - everything in Runtime.set.globals
+//  - everything in Runtime.blocks
+func dumpAll(a Arguments, depth int) reflect.Value {
+	var b bytes.Buffer
+	var vars VarMap
+
+	ctx := a.runtime.context
+	fmt.Fprintln(&b, "Context:")
+	fmt.Fprintf(&b, "\t%s %#v\n", ctx.Type(), ctx)
+
+	dumpScopeVars(&b, a.runtime.scope, 0)
+	dumpScopeVarsToDepth(&b, a.runtime.parent, depth)
+
+	vars = a.runtime.set.globals
+	for i, name := range vars.SortedKeys() {
+		if i == 0 {
+			fmt.Fprintln(&b, "Globals:")
+		}
+		val := vars[name]
+		fmt.Fprintf(&b, "\t%s:=%#v // %s\n", name, val, val.Type())
+	}
+
+	blockKeys := a.runtime.scope.sortedBlocks()
+	fmt.Fprintln(&b, "Blocks:")
+	for _, k := range blockKeys {
+		block := a.runtime.blocks[k]
+		dumpBlock(&b, block)
+	}
+
+	return reflect.ValueOf(b.String())
+}
+
+// dumpScopeVarsToDepth prints all variables in the scope, and all parent scopes,
+// to the limit of maxDepth.
+func dumpScopeVarsToDepth(w io.Writer, scope *scope, maxDepth int) {
+	for i := 1; i <= maxDepth; i++ {
+		if scope == nil {
+			break // do not panic if something bad happens
+		}
+		dumpScopeVars(w, scope, i)
+		scope = scope.parent
+	}
+}
+
+// dumpScopeVars prints all variables in the scope.
+func dumpScopeVars(w io.Writer, scope *scope, lvl int) {
+	if scope == nil {
+		return // do not panic if something bad happens
+	}
+	tabs := strings.Repeat("\t", lvl) // delimiter
+	if lvl == 0 {
+		fmt.Fprint(w, "Variables in scope:\n")
+	} else {
+		fmt.Fprintf(w, "%sVariables in scope %d level(s) up:\n", tabs, lvl)
+	}
+	vars := scope.variables
+	for _, k := range vars.SortedKeys() {
+		fmt.Fprintf(w, "\t%s%s=%#v\n", tabs, k, vars[k])
```

Comment:
Here, let's indent only by one tab, so the different scope levels are listed one after the other, but not nested.

Reply:
4ba85eaabc9dadf230b2778ef6d6c09839d67a28